contrib/target

```diff
@@ -87,13 +100,53 @@ def main():
     target_iqn = args.iqn_base + "." + host_name + "." + args.target_name
     target_dir = os.path.join(args.root_dir, args.target_name)
 
+    if args.portal:
+        # Duplicate portals are user error.
+        if len(args.portal) != len(set(args.portal)):
+            raise ValueError(f"Duplicate portals: {args.portal}")
+    else:
+        # This works only if all other targets use the default portal. Once you
+        # add a target using non default portal you must specify the portal for
+        # new targets. If you don't care about multiple portals, this default
+        # make it easy to use.
+        args.portal = [DEFAULT_PORTAL]
 
     if args.action == "create":
         create_target(args, target_iqn, target_dir, exists=args.exists)
     else:
         delete_target(args, target_iqn, target_dir)
 
 
+def portal(s):
+    _, port = DEFAULT_PORTAL
+    if s[0] == "[":
+        # IPv6: "[address]" or "[address]:port"
+        end = s.find("]")
+        if end == -1:
+            raise ValueError(s)
+        address = s[1:end]
+        if end + 1 < len(s):
+            if s[end+1] != ":":
+                raise ValueError(s)
+            port = int(s[end+2:])
+    else:
+        # IPv4: address or address:port
+        if ":" in s:
+            address, port = s.split(":", 1)
+            port = int(port)
+        else:
+            address = s
+    return address, port
```

Comment:
> I'm not sure to what extent do you want to check the validity of the input here (portal function), it could be fine as it is, but this can break easily.
> 
> A couple examples that would fail on trying to cast the port to integer:
> 
> * `[fe80::5054:ff:fede:9c26]:` -> out of bounds when trying to retrieve the ipv6 port

Yes, but this is fine since `[fe80::5054:ff:fede:9c26]:[1000:]` returns an empty string, and int('') rasies Value error, so we get the expected result:

```
$ contrib/target create 01 --portal "[::]:"
usage: target [-h] [-s LUN_SIZE] [-n LUN_COUNT] [-r ROOT_DIR] [-i IQN_BASE] [--cache] [--exists] [--portal PORTAL] {create,delete} target_name
target: error: argument --portal: invalid portal value: '[::]:'
```

> * `192.168.1.1:321:214` -> splits on the first `:`, the rest is not convertible to integer

Yes, so int() raises and we get:

```
$ contrib/target create 01 --portal "1.2.3.4:2000:3000"
usage: target [-h] [-s LUN_SIZE] [-n LUN_COUNT] [-r ROOT_DIR] [-i IQN_BASE] [--cache] [--exists] [--portal PORTAL] {create,delete} target_name
target: error: argument --portal: invalid portal value: '1.2.3.4:2000:3000'

```
 
> Also, the validity of the address is not checked. Examples for ipv4: the number of dots is 3, if the values between dots fit in 8 bits, etc.

Right, this will break in targetcli when using the bad address.
Validating ipv4 and specially ipv6 address is too much work for
this tool, so I leave it for targetcli.

Reply:
Regarding validating the addresses, maybe the best thing is to use socket.getaddrinfo():

It handles valid addresses:
```
>>> for addr in ["0", "1", "1.2", "1.2.3", "1.2.3.4", "::", "::1", "a:b:c:d::"]:
...     print(addr, "->", socket.getaddrinfo(addr, 3260, type=socket.SOCK_STREAM))
... 
0 -> [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '0', ('0.0.0.0', 3260))]
1 -> [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '1', ('0.0.0.1', 3260))]
1.2 -> [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '1.2', ('1.0.0.2', 3260))]
1.2.3 -> [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '1.2.3', ('1.2.0.3', 3260))]
1.2.3.4 -> [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '1.2.3.4', ('1.2.3.4', 3260))]
:: -> [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '::', ('::', 3260, 0, 0))]
::1 -> [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '::1', ('::1', 3260, 0, 0))]
a:b:c:d:: -> [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, 'a:b:c:d::', ('a:b:c:d::', 3260, 0, 0))]
```

And fail with invalid addresses:
```
>>> for addr in ["", "a.b.c.d", "256.0.0.0", "x:x::"]:
...     try:
...             socket.getaddrinfo(addr, 3260, type=socket.SOCK_STREAM)
...     except Exception as e:
...             print(addr, "->", e)
... 
 -> [Errno -2] Name does not resolve
a.b.c.d -> [Errno -2] Name does not resolve
256.0.0.0 -> [Errno -2] Name does not resolve
x:x:: -> [Errno -3] Try again
```

Although bad address results in trying to resolve the address, which may be slow.

Reply:
@almusil should we validate addresses in this developer tool? what is the best way to do this?

Reply:
What about [ipaddress](https://docs.python.org/3/library/ipaddress.html)? It does not try to resolve anything, and raises a ValueError.

```
>>> for addr in ["0", "1", "1.2", "1.2.3", "1.2.3.4", "::", "::1", "a:b:c:d::"]:
...     try:
...         ipaddress.ip_address(addr)
...     except ValueError:
...         print(f'Invalid address {addr}')
... 
Invalid address 0
Invalid address 1
Invalid address 1.2
Invalid address 1.2.3
IPv4Address('1.2.3.4')
IPv6Address('::')
IPv6Address('::1')
IPv6Address('a:b:c:d::')
```

Reply:
`getaddrinfo` tries to resolve what it thinks is a hostname instead of IP so yeah that can take some time. Using `ipaddress` module is probably a way to go.

Reply:
But ipaddress is wrong:
```
Invalid address 0
Invalid address 1
Invalid address 1.2
Invalid address 1.2.3
```

all these are working addresses, you can use them to listen or connect.

Reply:
But it is unlikely that someone will try to use address like "0" as a shortcut for "0.0.0.0", so using ipaddress sounds good.